interfaces/seccomp/backend_test.go

```diff
@@ -0,0 +1,257 @@
+	s.repo = interfaces.NewRepository()
+	s.iface = &interfaces.TestInterface{InterfaceName: "iface"}
+	err = s.repo.AddInterface(s.iface)
+	c.Assert(err, IsNil)
+}
+
+func (s *backendSuite) TearDownTest(c *C) {
+	dirs.SetRootDir("/")
+}
+
+// Tests for Configure() and Deconfigure()
+const sambaYamlV1 = `
+name: samba
+version: 1
+developer: acme
+apps:
+    smbd:
+slots:
+    iface:
+`
+const sambaYamlV1WithNmbd = `
+name: samba
+version: 1
+developer: acme
+apps:
+    smbd:
+    nmbd:
+slots:
+    iface:
+`
+const sambaYamlV2 = `
+name: samba
+version: 2
+developer: acme
+apps:
+    smbd:
+slots:
+    iface:
+`
+
+func (s *backendSuite) TestInstallingSnapWritesProfiles(c *C) {
+	developerMode := false
+	s.installSnap(c, developerMode, sambaYamlV1)
+	profile := filepath.Join(dirs.SnapSeccompDir, "snap.samba.smbd")
+	// file called "snap.sambda.smbd" was created
+	_, err := os.Stat(profile)
+	c.Check(err, IsNil)
+}
+
+func (s *backendSuite) TestRemovingSnapRemovesProfiles(c *C) {
+	for _, developerMode := range []bool{true, false} {
+		snapInfo := s.installSnap(c, developerMode, sambaYamlV1)
+		s.removeSnap(c, snapInfo)
+		profile := filepath.Join(dirs.SnapSeccompDir, "snap.samba.smbd")
+		// file called "snap.sambda.smbd" was removed
+		_, err := os.Stat(profile)
+		c.Check(os.IsNotExist(err), Equals, true)
+	}
+}
+
+func (s *backendSuite) TestUpdatingSnapToOneWithMoreApps(c *C) {
+	for _, developerMode := range []bool{true, false} {
+		snapInfo := s.installSnap(c, developerMode, sambaYamlV1)
+		snapInfo = s.updateSnap(c, snapInfo, developerMode, sambaYamlV1WithNmbd)
+		profile := filepath.Join(dirs.SnapSeccompDir, "snap.samba.nmbd")
+		// file called "snap.sambda.nmbd" was created
+		_, err := os.Stat(profile)
+		c.Check(err, IsNil)
+		s.removeSnap(c, snapInfo)
+	}
+}
+
+func (s *backendSuite) TestUpdatingSnapToOneWithFewerApps(c *C) {
+	for _, developerMode := range []bool{true, false} {
+		snapInfo := s.installSnap(c, developerMode, sambaYamlV1WithNmbd)
+		snapInfo = s.updateSnap(c, snapInfo, developerMode, sambaYamlV1)
+		profile := filepath.Join(dirs.SnapSeccompDir, "snap.samba.nmbd")
+		// file called "snap.sambda.nmbd" was removed
+		_, err := os.Stat(profile)
+		c.Check(os.IsNotExist(err), Equals, true)
+		s.removeSnap(c, snapInfo)
+	}
+}
+
+func (s *backendSuite) TestRealDefaultTemplateIsNormallyUsed(c *C) {
+	snapInfo, err := snap.InfoFromSnapYaml([]byte(sambaYamlV1))
+	c.Assert(err, IsNil)
+	// NOTE: we don't call seccomp.MockTemplate()
+	err = s.backend.Configure(snapInfo, false, s.repo)
+	c.Assert(err, IsNil)
+	profile := filepath.Join(dirs.SnapSeccompDir, "snap.samba.smbd")
+	data, err := ioutil.ReadFile(profile)
+	c.Assert(err, IsNil)
+	for _, line := range []string{
+		// NOTE: a few randomly picked lines from the real profile.  Comments
+		// and empty lines are avoided as those can be discarded in the future.
+		"deny kexec_load\n",
+		"deny create_module\n",
+		"symlink\n",
+		"socket\n",
```

Comment:
Note, the 'deny' rules are actually not supported by the launcher. The deny rules were used by sc-filtergen (15.04) and snappy policyGen (16.04) to strip out the specified syscall. Ie, if you had this policy:

deny kexec_load
kexec_load

then the profile generator would create (ie, commenting out the first *and* the second kexec_load):
# EXPLICITLY DENIED: kexec_load
# EXPLICITLY DENIED: kexec_load

It appears that at least as of today, 16.04 is incorrectly creating seccomp policy with 'deny' rules which are only working by accident (the launcher will see 'deny kexec_load' and allow 'deny' as a syscall, which happens to not exist).

I've been thinking about moving deny rules from the preprocessor (ie, snappy) and making them understood by the launcher. Seeing as snappy is creating invalid policy atm, I think I'll do that sonner than later.

Don't use socket since it will be changing with seccomp arg filtering that is close to landing.

Here is the list I suggest you use:
deny init_module
open
getuid